Getting started kafka connect

laidan6000 opened the pull request:
### Type of change

- Documentation

### Description

This pull request makes changes to the _Kafka Connect_ chapter of the _Getting Started_ section based on feedback received.

- Added a missing prerequisite the procedure for Creating a Docker image from a base image
- Changed titles of procedures for clarity
- Edited throughout for style, terminology and consistency.

### Checklist

_Please go through this checklist and make sure all applicable tasks have been done_

- [ ] Write tests
- [ ] Make sure all tests pass
- [ y] Update documentation
- [ ] Check RBAC rights for Kubernetes / OpenShift roles
- [ ] Try your changes from Pod inside your Kubernetes and OpenShift cluster, not just locally
- [ ] Reference relevant issue(s) and close them after merging
- [ ] Update CHANGELOG.md

ncbaratta commented on documentation/book/assembly-kafka-connect.adoc:
```diff
@@ -9,15 +9,13 @@
+
+- A source connector, named `FileStreamSourceConnector`.
+
+- A sink connector, named `FileStreamSyncConnector`.
```
Did you mean 'sync' or 'sink'

ppatierno replied:
It's sink ;)

ncbaratta commented on documentation/book/proc-creating-new-image-from-base.adoc:
```diff
@@ -3,17 +3,22 @@
+.Prerequisites
+
+ifdef::Kubernetes[]
+* xref:deploying-cluster-operator-kubernetes-{context}[Deploying the Cluster Operator to {KubernetesName}]
```
I believe you want to replace {context} here with whatever the context was on the page you're linking to

ncbaratta commented on documentation/book/proc-creating-new-image-from-base.adoc:
```diff
@@ -3,17 +3,22 @@
+* xref:deploying-cluster-operator-kubernetes-{context}[Deploying the Cluster Operator to {KubernetesName}]
+endif::Kubernetes[]
+
+* xref:deploying-cluster-operator-openshift-{context}[Deploying the Cluster Operator to {OpenShiftName}]
```
I believe you want to replace {context} here with whatever the context was on the page you're linking to

ncbaratta commented on documentation/book/proc-creating-new-image-from-base.adoc:
```diff
@@ -3,17 +3,22 @@
 
 .Procedure
 
+To create a custom image and add your connector to the `/opt/kafka/plugins` directory:
```
We don't want to have an intro sentence in the procedure, so you can include this up above the prereqs or turn this a bullet somehow

ncbaratta commented on documentation/book/proc-deploying-kafka-connect-kubernetes.adoc:
```diff
@@ -3,20 +3,24 @@
 .Prerequisites
 
-* Before deploying Kafka Connect, the Cluster Operator must be deployed.
+* xref:deploying-cluster-operator-kubernetes-{context}[Deploying the Cluster Operator to {KubernetesName}]
```
I believe you want to replace {context} here with whatever the context was on the page you're linking to

ncbaratta commented on documentation/book/proc-deploying-kafka-connect-openshift.adoc:
```diff
@@ -3,22 +3,23 @@
 .Prerequisites
 
-* Before deploying Kafka Connect, the Cluster Operator must be deployed.
+* xref:deploying-cluster-operator-openshift-{context}[Deploying the Cluster Operator to {OpenShiftName}]
```
I believe you want to replace {context} here with whatever the context was on the page you're linking to

ncbaratta commented on documentation/book/proc-using-openshift-builds-create-image.adoc:
```diff
@@ -4,28 +4,30 @@
 
 [id='using-openshift-s2i-create-image-{context}']
-= Using {OpenShiftName} builds and S2I to create new images
+= Creating a container image using OpenShift builds
```
Replace OpenShift with {OpenShiftName}

ncbaratta commented on documentation/book/proc-using-openshift-builds-create-image.adoc:
```diff
@@ -4,28 +4,30 @@
 .Procedure
 
-. Create a Kafka Connect S2I cluster from the command-line
+To create a new container image using an {OpenShiftName} build:
```
We don't want an intro sentence in the procedure itself, only numbered items.

scholzj replied:
This is not what it does. What you do in this step is you deploy Kafka Connect with S2I support.

ncbaratta commented on documentation/book/assembly-kafka-connect.adoc:
```diff
@@ -9,15 +9,13 @@
-xref:assembly-deployment-configuration-kafka-connect-s2i-str[Kafka Connect cluster with Source2Image support].
+link:https://kafka.apache.org/documentation/#connect[Kafka Connect^], a core component of Apache Kafka, is a tool for streaming high volumes of data into and out of your Kafka cluster. It provides a framework for building connectors (a type of plug-in) for your Kafka cluster, a REST API for managing connectors, and more. Kafka Connect includes two built-in connectors, which are both available for ...
+
+- A source connector, named `FileStreamSourceConnector`.
```
If this is a bulleted list you want to use * instead of -

ppatierno commented on documentation/book/assembly-kafka-connect.adoc:
```diff
@@ -9,15 +9,13 @@
-xref:assembly-deployment-configuration-kafka-connect-s2i-str[Kafka Connect cluster with Source2Image support].
+link:https://kafka.apache.org/documentation/#connect[Kafka Connect^], a core component of Apache Kafka, is a tool for streaming high volumes of data into and out of your Kafka cluster. It provides a framework for building connectors (a type of plug-in) for your Kafka cluster, a REST API for managing connectors, and more. Kafka Connect includes two built-in connectors, which are both available for ...
+
+- A source connector, named `FileStreamSourceConnector`.
```
Maybe is it better saying "a file source connector" ?

laidan6000 replied:
Decided to present the information in a table rather than a bullet list. The header rows are **File Connector** and **Description**.

ppatierno commented on documentation/book/assembly-kafka-connect.adoc:
```diff
@@ -9,15 +9,13 @@
+
+- A source connector, named `FileStreamSourceConnector`.
+
+- A sink connector, named `FileStreamSyncConnector`.
```
"a file sink connector" ?

ppatierno commented on documentation/book/assembly-kafka-connect.adoc:
```diff
@@ -9,15 +9,13 @@
-xref:assembly-deployment-configuration-kafka-connect-str[Kafka Connect cluster configuration]
-and
-xref:assembly-deployment-configuration-kafka-connect-s2i-str[Kafka Connect cluster with Source2Image support].
+link:https://kafka.apache.org/documentation/#connect[Kafka Connect^], a core component of Apache Kafka, is a tool for streaming high volumes of data into and out of your Kafka cluster. It provides a framework for building connectors (a type of plug-in) for your Kafka cluster, a REST API for managing connectors, and more. Kafka Connect includes two built-in connectors, which are both available for ...
```
you can paraphrase better than me but what about "...two built-in connectors for moving files content in and out a of the Kafka cluster..."

laidan6000 replied:
The connectors are tough to explain, even though I used them in my Kafka Essentials training. Does this make sense?

These connectors transfer data into and out of a Kafka cluster by using external files for sending and receiving messages.

ppatierno replied:
With this sentence it seems to me that files are used as a "tool" for transferring messages between Kafka and a different other system. Instead, you are getting data from file to kafka (source) or putting data from Kafka to a file (sink).

laidan6000 replied:
> Kafka Connect includes the following built-in connectors for moving file-based data into and out of your Kafka cluster.

(Table)

laidan6000 commented on documentation/book/proc-using-openshift-builds-create-image.adoc:
```diff
@@ -4,28 +4,30 @@
-{OpenShiftName} supports link:https://docs.openshift.org/3.9/dev_guide/builds/index.html[builds^], which can be used together with the link:https://docs.openshift.org/3.9/creating_images/s2i.html#creating-images-s2i[Source-to-Image (S2I)^] framework to create new container images.
-An {OpenShiftName} build takes a builder image with S2I support together with the connector plugin binaries provided by the user and uses them to build a new container image.
-The newly created container image is stored in {OpenShiftName}'s local container image repository and can be used in deployments.
+You can use {OpenShiftName} link:https://docs.openshift.org/3.9/dev_guide/builds/index.html[builds^] and the  link:https://docs.openshift.org/3.9/creating_images/s2i.html#creating-images-s2i[Source-to-Image (S2I)^] framework to create new container images. In an {OpenShiftName} build, a new container image is built from the following components:
```
OK. I've reverted to the existing sentence. The bullets were an attempt to break up the sentence to make it easier to digest, but I must have changed the meaning somehow.

laidan6000 commented:
Thanks for the detailed review, @ncbaratta, @ppatierno and @scholzj. I've addressed your comments and pushed the changes. 

Please have a second pass and approve if OK. Unfortunately, I can't deploy the modified version to a staging site at the moment.

ncbaratta commented:
Travis failing so you should look into that - reviewing content now.

ncbaratta commented on documentation/book/assembly-using-kafka-connect-with-plugins.adoc:
```diff
@@ -6,16 +6,13 @@
+= Extending Kafka Connect with plug-ins
 
-{ProductName} container images for Kafka Connect contain, by default, only the `FileStreamSinkConnector` and `FileStreamSourceConnector` connectors which are part of Apache Kafka.
+The {ProductName} container images for Kafka Connect include the two built-in connectors: `FileStreamSourceConnector` and `FileStreamSinkConnector`. You can add your own connectors by using one of the following methods:
```
two shouldn't be spelled out.

ncbaratta replied:
Wait ... yes it should. I always screw these up.  It's numbers over nine that are not spelled out. ignore me @laidan6000

laidan6000 replied:
Ha ha, OK ;)

ncbaratta commented on documentation/book/proc-creating-new-image-from-base.adoc:
```diff
@@ -24,10 +30,19 @@ COPY ./_my-plugins_/ /opt/kafka/plugins/
+. Upload your custom image to your container registry.
 
-. Set the `KafkaConnect.spec.image` property of the KafkaConnect custom resource or the `STRIMZI_DEFAULT_KAFKA_CONNECT_IMAGE` variable to point to the new container image.
+. Perform one of the following tasks:
```
Why are we offering two options? Do they achieve the same thing? If so I say we should be more prescriptive.

laidan6000 replied:
I'd need some input from @ppatierno on this. ;)

ppatierno replied:
Well ... the Cluster Operator (CO) YAML file has DEFAULT for every image that we use for the different containers (Kafka, Zookeeper, Kafka Connect and so on). It means that the CO will use that DEFAULT image if the user doesn't set anything in the `image` field for the specific resource (i.e. if `KafkaConnect.spec.image` is not set, the CO will create a Kafka Connect instance using the `STRIMZI_DEFAULT_KAFKA_CONNECT_IMAGE`). As opposite, the `image` property, if set, override the DEFAULT one.

ppatierno commented:
I think that failing is just related to flaky tests. Just try to respin the Travis CI build.

laidan6000 commented:
Will do. I can't find any AsciiDoctor errors in the log.

scholzj commented on documentation/book/assembly-kafka-connect.adoc:
```diff
@@ -9,15 +9,18 @@
-xref:assembly-deployment-configuration-kafka-connect-str[Kafka Connect cluster configuration]
-and
-xref:assembly-deployment-configuration-kafka-connect-s2i-str[Kafka Connect cluster with Source2Image support].
+link:https://kafka.apache.org/documentation/#connect[Kafka Connect^], a core component of Apache Kafka, is a tool for streaming high volumes of data into and out of your Kafka cluster. It provides a framework for building connectors (types of plug-ins) for your Kafka cluster, a REST API for managing connectors, and more. It includes two built-in connectors:
```
_streaming high volume_ sounds strange. Wouldn't just _transfering data into and out of ..._ be better?

laidan6000 replied:
I've reconciled this with the description we have in the RHEL documentation. "It provides a framework for moving large amounts of data into and out of your cluster while maintaining scalability and reliability."

scholzj replied:
Sounds good. Thanks.

scholzj commented on documentation/book/assembly-kafka-connect.adoc:
```diff
@@ -9,15 +9,18 @@
-xref:assembly-deployment-configuration-kafka-connect-str[Kafka Connect cluster configuration]
-and
-xref:assembly-deployment-configuration-kafka-connect-s2i-str[Kafka Connect cluster with Source2Image support].
+link:https://kafka.apache.org/documentation/#connect[Kafka Connect^], a core component of Apache Kafka, is a tool for streaming high volumes of data into and out of your Kafka cluster. It provides a framework for building connectors (types of plug-ins) for your Kafka cluster, a REST API for managing connectors, and more. It includes two built-in connectors:
```
From my POV, connector is instance, connector plugin is the JAR file. So it is framework for building connector plugins and it is framework for running connectors. (It is both in a sense)

scholzj commented on documentation/book/assembly-kafka-connect.adoc:
```diff
@@ -9,15 +9,18 @@
+
+|===
+| *Connector* | *Description*
+| `FileStreamSourceConnector` | Transfers data to your Kafka cluster from an external file (the source).
```
I'm not sure what is _external file_. It should be just _file_.

scholzj commented on documentation/book/assembly-kafka-connect.adoc:
```diff
@@ -9,15 +9,18 @@
+In {ProductName}, you can use the Cluster Operator to deploy a Kafka Connect cluster to your {ProductPlatformName} cluster. Within your {ProductPlatformName} cluster:
+
+* A Kafka Connect cluster is implemented as a `Deployment` with a configurable number of workers.
+* The Kafka Connect REST API is available on port 8083, as the `kafka-connect` service.
```
The service is not named 'kafka-connect'. It is named something like '<connect-cluster-name>-connect-api' or something like that.

laidan6000 replied:
@scholzj - Do you know the exact name?

scholzj replied:
`<connect-cluster-name>-connect-api` is the right name.

scholzj commented on documentation/book/assembly-kafka-connect.adoc:
```diff
@@ -9,15 +9,18 @@
+
+In {ProductName}, you can use the Cluster Operator to deploy a Kafka Connect cluster to your {ProductPlatformName} cluster. Within your {ProductPlatformName} cluster:
+
+* A Kafka Connect cluster is implemented as a `Deployment` with a configurable number of workers.
```
This list isn't valid for S2I enabled Connect deployment. I think you need to mention it here as well.

scholzj commented on documentation/book/assembly-kafka-connect.adoc:
```diff
@@ -9,15 +9,18 @@
+| `FileStreamSinkConnector` | Transfers data from your Kafka cluster to an external file (the sink).
+|===
+
+In {ProductName}, you can use the Cluster Operator to deploy a Kafka Connect cluster to your {ProductPlatformName} cluster. Within your {ProductPlatformName} cluster:
```
_Within your {ProductPlatformName} cluster:_ doesn't sound like an actual sentence. I think it is missing some verb either here or in the bullet points.

laidan6000 replied:
Reformatted as a complete paragraph.

scholzj commented on documentation/book/assembly-using-kafka-connect-with-plugins.adoc:
```diff
@@ -6,16 +6,13 @@
-
-* Using a custom Docker image
-* Using the {OpenShiftName} build system with the {ProductName} S2I
+* Create a Docker image from the Kafka Connect base image.
```
You should use _container image_ everywhere.

scholzj commented on documentation/book/proc-creating-new-image-from-base.adoc:
```diff
@@ -3,14 +3,20 @@
+A container image for running Kafka Connect using {ProductName} is available on {DockerRepository} as `{DockerKafkaConnect}`. You can use this as a base image for creating your own custom image with additional plug-ins. 
 
-The following procedure describes the process for creating such a custom image and adding the connector plugins to the the `/opt/kafka/plugins` directory.
+At startup, Kafka Connect loads any third-party plug-ins, including connectors, contained in the `/opt/kafka/plugins` directory. This procedure describes how to create a custom image and add your connector to this directory. 
```
_including connectors_ sounds a bit strange. Basically, the plugins are connectors.

laidan6000 replied:
Agreed. I've simplified the terminology to just say "connector plug-ins" as you suggested earlier.

scholzj commented on documentation/book/proc-creating-new-image-from-base.adoc:
```diff
@@ -3,14 +3,20 @@
+A container image for running Kafka Connect using {ProductName} is available on {DockerRepository} as `{DockerKafkaConnect}`. You can use this as a base image for creating your own custom image with additional plug-ins. 
 
-The following procedure describes the process for creating such a custom image and adding the connector plugins to the the `/opt/kafka/plugins` directory.
+At startup, Kafka Connect loads any third-party plug-ins, including connectors, contained in the `/opt/kafka/plugins` directory. This procedure describes how to create a custom image and add your connector to this directory. 
```
This should not say that _Kafka Connect does this_. AMQ Streams Kafka Connect does this.

laidan6000 replied:
Tough to explain! I've changed as follows:

> At startup, the AMQ Streams version of Kafka Connect loads any third-party connector plug-ins contained in the `/opt/kafka/plugins` directory.

scholzj replied:
That sounds good. Thanks.

scholzj commented on documentation/book/proc-creating-new-image-from-base.adoc:
```diff
@@ -24,10 +30,19 @@ COPY ./_my-plugins_/ /opt/kafka/plugins/
-. Build the container image and upload it to the appropriate container image repository.
+. Build the container image.
+
+. Upload your custom image to your container registry.
```
Upload => Push

scholzj commented on documentation/book/proc-creating-new-image-from-base.adoc:
```diff
@@ -24,10 +30,19 @@ COPY ./_my-plugins_/ /opt/kafka/plugins/
-. Set the `KafkaConnect.spec.image` property of the KafkaConnect custom resource or the `STRIMZI_DEFAULT_KAFKA_CONNECT_IMAGE` variable to point to the new container image.
+. Perform one of the following tasks:
+
+* Edit the `KafkaConnect.spec.image` property of the `KafkaConnect` custom resource to point to the new container image.
```
Should we add an example?

laidan6000 replied:
Added an example.

scholzj commented on documentation/book/proc-creating-new-image-from-base.adoc:
```diff
@@ -24,10 +30,19 @@ COPY ./_my-plugins_/ /opt/kafka/plugins/
+
+* For more information on the `KafkaConnect.spec.image property`, see xref:assembly-configuring-container-images-deployment-configuration-kafka-connect[].
+
+
```
Do we need two empty lines here?

scholzj commented on documentation/book/proc-deploying-kafka-connect-kubernetes.adoc:
```diff
@@ -3,20 +3,24 @@
 
 [id='deploying-kafka-connect-kubernetes-{context}']
-= Deploying Kafka Connect to {KubernetesName}
+= Deploying Kafka Connect to your {KubernetesName} cluster
```
Is there any reason why do we have two procedures ... one for Kube and one for OpenShift? Most other procedures are one procedure for both.

laidan6000 replied:
I'm not sure why this is the case here.

laidan6000 commented on documentation/book/assembly-using-kafka-connect-with-plugins.adoc:
```diff
@@ -6,16 +6,13 @@
+= Extending Kafka Connect with plug-ins
 
-{ProductName} container images for Kafka Connect contain, by default, only the `FileStreamSinkConnector` and `FileStreamSourceConnector` connectors which are part of Apache Kafka.
+At startup, Kafka Connect loads any third-party plug-ins, including connectors, contained in the `/opt/kafka/plugins` directory. You can add your own connectors to this directory in one of two ways:
```
I moved this information into the introduction to the relevant procedure, "Creating a Docker image from the Kafka Connect base image".

laidan6000 commented on documentation/book/proc-deploying-kafka-connect-kubernetes.adoc:
```diff
@@ -3,20 +3,24 @@
-= Deploying Kafka Connect to {KubernetesName}
+= Deploying Kafka Connect to your {KubernetesName} cluster
+
+To allow you to build connectors for your Kafka cluster, you can deploy a Kafka Connect cluster to your {KubernetesName} cluster by using the Cluster Operator.
```
The first clause has been removed from that sentence.
